Switch age fix

inspectredc opened the pull request:
This PR addresses the bugs for the switch age cheat, noted in issues #2562 and #2013. The latter's solution can be used for save states too but I thought it would be best to fix those in a separate PR.


https://user-images.githubusercontent.com/78732756/230623938-ec992892-65ac-4929-b580-61b9ddee59db.mp4


<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/637377145.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/637377147.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/637377148.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/637377149.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/637377150.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/637377151.zip)
<!--- section:artifacts:end -->

inspectredc committed:
specific scenarios for switch age button, non rando

inspectredc committed:
Working fix for vanilla swordless change age

inspectredc committed:
Fix now works for randomiser

inspectredc committed:
Fixed room load on change age transition

inspectredc committed:
Removed potentially unnecessary function call

Need to investigate this more, maybe wait for more room documentation.

Archez commented on soh/src/code/z_parameter.c:
```diff
@@ -1533,8 +1541,18 @@ void Inventory_SwapAgeEquipment(void) {
+            }
+        }
+    }
+    CVarSetInteger("gSwitchAge", 0);
```
I wonder if this can be moved to the second half of the switch age hook (in the `if (warped ...)` check, so the logic for setting/unsetting is contained in the hook itself.

inspectredc replied:
yea thats probably best, I initially had issues with it setting back to 0 early but should work now with the extra '!warped' condition added

inspectredc replied:
yep still works as expected

Archez commented on soh/src/code/z_parameter.c:
```diff
@@ -1411,24 +1411,31 @@ void Inventory_SwapAgeEquipment(void) {
             } else {
-                gSaveContext.childEquips.buttonItems[i] = ITEM_SWORD_KOKIRI;
+                if (CVarGetInteger("gSwitchAge", 0) && 
+                    gSaveContext.infTable[29]) {
```
I know `infTable[29]` is strictly only used for the swordless flag, so this check works, but maybe it should be more explicit like `(gSaveContext.infTable[29] & 1)`  or `gSaveContext.infTable[29] != 0`

inspectredc replied:
alright yea, former makes the most sense for me if done for clarity

inspectredc committed:
Suggested Edits

Moves setting cvar back into mods.cpp and makes clear the condition expected for the no sword flag

Archez reviewed (approved):
Awesome looks good!
Glad those room change methods were pretty straight forward to use. When I originally implemented Boss shuffle, I needed a way to fix broken entrances and I stumbled upon these methods. My first thought was that they could be used to help improve save states. Excited to see a follow up PR for that if possible 🙂
